[CSS Container Queries][Style queries] Update style correctly for non-inherited custom property mutations

anttijk opened the pull request:
#### 801286aa3dcb9720da3f98e6d25ce5e1ab5b79bb
<pre>
[CSS Container Queries][Style queries] Update style correctly for non-inherited custom property mutations
<a href="https://bugs.webkit.org/show_bug.cgi?id=270110">https://bugs.webkit.org/show_bug.cgi?id=270110</a>
<a href="https://rdar.apple.com/123645196">rdar://123645196</a>

Reviewed by Alan Baradlay.

Mutating a custom property referenced in a style query needs to re-evaluate descendant styles or potential
style containers. This needs to happen also when the custom property is non-inherited.

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/container-queries/custom-property-style-query-change-expected.txt:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::customPropertyValueEqual const):

Add a helper.

* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::RuleSet::containerQueryRules const):
* Source/WebCore/style/RuleSet.h:
* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ScopeRuleSets::collectFeatures const):
(WebCore::Style::ScopeRuleSets::customPropertyNamesInStyleContainerQueries const):

Collect all the property names used in style queries.

* Source/WebCore/style/StyleScopeRuleSets.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::computeDescendantsToResolve const):

Check if we have property dependent style queries and if so see if their value has changed.
If it has we need to re-resolve the descendants.

(WebCore::Style::TreeResolver::resolveElement):
(WebCore::Style::TreeResolver::computeDescendantsToResolve): Deleted.
* Source/WebCore/style/StyleTreeResolver.h:

Canonical link: <a href="https://commits.webkit.org/275445@main">https://commits.webkit.org/275445@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e90cdd1633e1b917e3af8f30a997392018f0f91

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41694 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20708 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44075 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44266 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37787 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44001 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23835 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18038 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34454 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42268 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17631 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35904 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15124 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15324 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36925 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45653 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37875 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37246 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40996 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16509 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13549 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39421 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18128 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9379 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18185 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17772 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/style/StyleScopeRuleSets.cpp:
```diff
@@ -344,6 +346,30 @@ const Vector<InvalidationRuleSet>* ScopeRuleSets::hasPseudoClassInvalidationRule
+        auto collectPropertyNames = [&](auto* ruleSet) {
+            if (!ruleSet)
+                return;
+            for (auto query : ruleSet->containerQueryRules()) {
```
`const auto& query`

mdubet replied:
(we don't need to increment the refcounter for a refcounted member variable IIRC)

mdubet replied:
This hasn't been changed before the merge 😞

anttijk replied:
const adds nothing here (rules are already const). Using references is less safe and since this is not performance critical there is no particular need to avoid churn.

At least my policy for cosmetic suggestions is that following them is at the discretion of the patch author.

mdubet commented on Source/WebCore/style/StyleScopeRuleSets.cpp:
```diff
@@ -277,6 +277,8 @@ void ScopeRuleSets::collectFeatures() const
     m_pseudoClassInvalidationRuleSets.clear();
     m_hasPseudoClassInvalidationRuleSets.clear();
 
+    m_customPropertyNamesInStyleContainerQueries = std::nullopt;
```
Not that important here because there is no path forward to generalise this, but uniform initialisation should be preferred `{ }`

mdubet commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -3465,6 +3465,20 @@ const CSSCustomPropertyValue* RenderStyle::customPropertyValue(const AtomString&
 
+bool RenderStyle::customPropertyValueEqual(const RenderStyle& other, const AtomString& name) const
+{
+    if (&nonInheritedCustomProperties() == &other.nonInheritedCustomProperties() && &inheritedCustomProperties() == &other.inheritedCustomProperties())
```
Is this fairly complex condition different from `RenderStyle::customPropertiesEqual()` ?

anttijk replied:
`customPropertiesEqual` tests for true equality (which may involve checking every single property). This just does the fast pointer equality test as a shortcut (custom property data is often shared).

anttijk committed:
[CSS Container Queries][Style queries] Update style correctly for non-inherited custom property mutations
https://bugs.webkit.org/show_bug.cgi?id=270110
rdar://123645196

Reviewed by Alan Baradlay.

Mutating a custom property referenced in a style query needs to re-evaluate descendant styles or potential
style containers. This needs to happen also when the custom property is non-inherited.

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/container-queries/custom-property-style-query-change-expected.txt:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::customPropertyValueEqual const):

Add a helper.

* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::RuleSet::containerQueryRules const):
* Source/WebCore/style/RuleSet.h:
* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ScopeRuleSets::collectFeatures const):
(WebCore::Style::ScopeRuleSets::customPropertyNamesInStyleContainerQueries const):

Collect all the property names used in style queries.

* Source/WebCore/style/StyleScopeRuleSets.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::computeDescendantsToResolve const):

Check if we have property dependent style queries and if so see if their value has changed.
If it has we need to re-resolve the descendants.

(WebCore::Style::TreeResolver::resolveElement):
(WebCore::Style::TreeResolver::computeDescendantsToResolve): Deleted.
* Source/WebCore/style/StyleTreeResolver.h:

Canonical link: https://commits.webkit.org/275445@main
